DOC: More on pandas comparison (#2875)
pydata/xarray: doc/faq.rst

-------------------------

pandas is a fantastic library for analysis of low-dimensional labelled data -

> if data can be sensibly described as "rows and columns", pandas is probably the

Reviewer: ```suggestion
if the data can be sensibly described as "rows and columns", pandas is probably the
```

right choice.  However, xarray has some distinct advantages: higher dimensionality

> and an explicit, production-ready API.

Reviewer 2: this sounds a bit as if the pandas API was not production-ready

Author: I'll soften to a comparative

Reviewer 3: I'm not sure either xarray or pandas are production ready :)

Higher dimensionality
~~~~~~~~~~~~~~~~~~~~~

Sometimes we want to use higher dimensional arrays, or arrays for 
which the names of dimensions have some semantic significance we want to retain.
Pandas gets a lot of things right, but many science, engineering and complex
analytics use cases need fully multi-dimensional data structures.

> For example, consider the images of a movie, represented as an array of pixels with 

Reviewer 3: I'm not sure the movie example is the most compelling comparison to pandas. I think it's pretty obvious that a DataFrame is a poor fit for a movie -- the obvious data structure to use would be a NumPy array.

Maybe something 3D that could fit into a Panel or multi-indexed DataFrame would be a better choice? Examples that come to mind:
- open/high/low/close stock data (2D array of shape time/company in an xarray.Dataset) -- this was the textbook example for `pandas.Panel`
- [inference data in ArviZ](https://arviz-devs.github.io/arviz/notebooks/InferenceDataCookbook.html) with dimensions like (chain, draw, ...)

Author: Yes, I agree, the movie example isn't great as a comparison to pandas (good for numpy)
Is there something that's more generally intuitive than finance or Bayesian stats? I think weather might be our best bet! Open to others

four dimensions: time, row, column and color; xarray makes working with this 
much easier than if it were stacked into two dimensions:

  - *Name not position*: The dimensions are better represented by their name 
    than their position. By allowing us to use their names, our code becomes 
    easier to write and much easier to read; 
    e.g. ``da.isel(time=0)`` rather than ``da[:, :, 0]``.

  - *Broadcasting*: Xarray can broadcast operations to an array, meaning it
    applies the operation similarly regardless of the number of dimensions. 
    This allows our code to be more abstract, requiring far fewer special cases. 

    For example, ``da - 1`` would make every pixel slightly darker, whether 

>     we're dealing with a single or multiple frames, color or B&W, etc. 

Reviewer 3: ```suggestion
    we're dealing with a single or multiple frames, color or grayscale, etc. 
```

>     Similarly ``da - da.mean('row', 'column')`` would normalize the values 

Reviewer 3: ```suggestion
    Similarly ``da - da.mean(dim=['row', 'column'])`` would normalize the values 
```

    based on the mean value in its width and height, again regardless of whether

>     we have a single or multiple frames, color or B&W, etc.

Reviewer 3: ```suggestion
    we have a single or multiple frames, color or grayscale, etc.
```

  - *Alignment*: Xarray can package arrays which share some, but not all, of
    their dimensions in a dataset and align the dimensions they do share. 
    For example, if we calculated the brightness of our movie images

>     using ``da.mean('row', 'column', 'color')`, we'd have an array with a time

Reviewer 3: ```suggestion
    using ``da.mean(dim=['row', 'column', 'color'])``, we'd have an array with a time
```

    dimension. When we add that to a dataset and select the first point in time,
    ``ds.isel(time=0)``, we'll receive both the frame and the first brightness
    value. 

    Even following best-pracitices around tidy tabular data, this approach is

>     very difficult to replicate using tabular data.

Reviewer: ```suggestion
    very difficult or at least unintuitive to replicate using tabular data.
```
